_posts/2022-01-18-four-favorite-firmware-debug-tools.md

```diff
@@ -0,0 +1,126 @@
+
+One last note: the handheld models of multimeter tend to be fairly low precision. You'll have a hard time measuring anything below 0.01 digits of voltage, current, or resistance. If you need a lower level of detail - say, for measuring precision shunt resistors - a benchtop six digit multimeter from Keithley or similar is a better bet. However, this level of precision is not typically required for garden variety embedded work. 
+
+{% include newsletter.html %}
+
+## Saleae Logic Analyzer 
+
+So, if you've got a system driving a lot of toggling IO, and you need to figure out what's going on, what are you gonna do? Your multimeter, as we've just said, isn't gonna cut it. 
+
+An oscilloscope could certainly do the job. You can easily evaluate whether an IO is toggling, and measure time-based parameters like duty cycle and rise time. However, being embedded people, we generally care about interchip communications: things like UART, I2C, and SPI transactions. You can view these on oscilloscopes, but they are painful to decode by hand. Almost all scope vendors offer decoder packages, but they are expensive, and generally not bundled together. You have to pay a few hundred dollars a pop to unlock UART, then SPI, then I2C, and so on and so forth for the smorgasbord of digital signals you need to monitor. On top of that, the UI for many scopes is clunky. Digital protocol decoders can be a pain to configure and steer using only the rudimentary knobs and buttons common to most scope UIs. Remote desktops aren't much better - these tend to be laggy and difficult to drive. 
+
+What you really need is a logic analyzer - one that plays nicely with your development machine. 
+
+Fortunately for you, me, and the rest of us embedded folks, this machine exists. 
+
+The Saleae Logic Pro16 is the best logic analyzer on the market for embedded systems.
+
+### Logic is the Beginning of Wisdom
```

Comment:
Maybe add a link to how this tool can be quickly automated with python?

https://support.saleae.com/saleae-api-and-sdk/socket-api#python-wrapper-automation-method-1

Reply:
Good note - I've never used the API, I'll read up and figure out a graceful way to include it.

Reply:
@noahp updated with some details on the Socket API - please review!

Reply:
looks good, ty!